stablehlo/transforms/StablehloProbeInstrumentation.cpp

```diff
@@ -0,0 +1,113 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2023 The StableHLO Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <string>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOPROBEINSTRUMENTATIONPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+class StablehloProbeInstrumentationPass
+    : public impl::StablehloProbeInstrumentationPassBase<
+          StablehloProbeInstrumentationPass> {
+ public:
+  explicit StablehloProbeInstrumentationPass() = default;
+  void runOnOperation() override;
+
+ private:
+  // Create a uniquely identifying probe ID for the given value being
+  // instrumented. Attempt to use named MLIR location data, otherwise use an
+  // increasing ID.
+  std::string getLocationNameOrUniqueId(Location location);
```

Comment:
> I.e. don't want models being written specially for probing which then ruins the general debug experience which uses file-line-column etc.

Completely agree. If we were to require this, it would not make the debugging experience work OOB and make it much less user friendly. Perhaps we could modify the pass description and/or `instrument.probe` op descriptions to indicate this (that location data is optional)? Or perhaps make it a pass option?

> I'd rather that debuginfo (location) doesn't have semantic meaning to ProbeOp. Can we just use the incrementing number `probe1`, `probe2`..

Initially, the thought behind this was to allow the instrumented data produced by the StableHLO reference interpreter to better align with other runtimes, as I imagine ultimately it will be possible to roundtrip StableHLO programs to other representations (such as TFLite, etc) given the portability goal of StableHLO.

For example, there was a [commit](https://github.com/tensorflow/tensorflow/commit/84101d7daef2b49add3da3418076547ad7333c3e) earlier this year in Tensorflow Lite to allow subgraphs to have their intermediate tensor values serialized to disk, for which tensor names could be ostensibly used for deriving unique identifiers (i.e. `probe_id`). My understanding is that the existing [flatbuffer_translate](https://github.com/tensorflow/tensorflow/blob/6f50f70c873d60a84baf35957190d8e8ccff5dd6/tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer/flex_op_with_f64.mlir#L1) tool will use these values to derive MLIR location data.

Suppose if there were a way to convert TFLite MLIR into StableHLO, and if the StableHLO probe instrumentation pass were to be run as part of a larger compiler pass pipeline, it may make it a lot harder to try and match the instrumentation locations without existing MLIR location data, as the compiler may shuffle, fuse or remove ops. I think the location data could also make it easier to discern where an instrumented value was produced in a model (i.e. a `convolution` vs `matmul`, etc). Ultimately, named MLIR location data wouldn't be required for instrumentation. It'd just be an option for maximizing usability with other runtimes.

Thoughts?

Reply:
I understand the use case, and given that this is a pass for interpreter testing it's probably OK to use debuginfo, as it will come out explicit in the resulting probed program.

Thinking about this a different way- Let's say I have infra set up to probe a model and have gathered probeinfo already, then the debuginfo (not the structure of the program) of the next export changes and I have a bunch of different names.

If you think this is a potential problem, perhaps we can hide `useDebugInfo` behind an opt-in pass option? If you think this isn't actually invasive then perhaps we're good as-is.